### ISSUE TITLE
vector_eq raises ERROR instead of returning FALSE. Is that intended behavior?

### ISSUE THREAD

**Issue by ergo70** [pgvector/pgvector#451]
Hi all,

vector_eq() calls CheckDims() first, raising an ERROR if the compared vectors have different sizes. Since pg_vector supports vectors of variable size, this causes problems. In this case, vector_eq() should return FALSE, not throw an ERROR. This also throws ERRORs when doing a VACUUM ANALYZE. Is that intended behavior?

best regards

Ernst-Georg

**Comment by ankane** [comment]
Hi @ergo70, I could see a case for changing this in 0.7.0. Can you explain more about what you're trying to do? Also, what errors are you seeing with `VACUUM ANALYZE`?

**Comment by ergo70** [comment]
Hi,

I just want to compare two vectors for equality:

select '[1,2,3]'::vector = '[1,2,3]'::vector; -> TRUE
select '[1,2,3]'::vector = '[3,4,5]'::vector; -> FALSE
select '[1,2]'::vector = '[3,4,5]'::vector; -> SQL Error [22000]: ERROR: different vector dimensions 2 and 3

BUT it should be FALSE, IMHO and not throw an ERROR.

PostgreSQL ARRAYs behave correct:

select '{1,2}'::int4[] = '{3,4,5}'::int4[]; -> FALSE

vacuum <my_vector_table_with_variable_size_vectors>; -> OK
vacuum analyze <my_vector_table_with_variable_size_vectors>; -> SQL Error [22000]: FEHLER: different vector dimensions 10 and 22
analyze <my_vector_table_with_variable_size_vectors>; -> SQL Error [22000]: FEHLER: different vector dimensions 10 and 22

The same happens every time autovacuum runs.

best regards,

Ernst-Georg

**Comment by ankane** [comment]
Pushed a fix for `ANALYZE` in the commit above, and will change the operators to not throw an error in 0.7.0. Thanks for reporting and the suggestion!

**Comment by hlinnaka** [comment]
This still fails:
```
postgres=# CREATE TABLE dimtest (id int, dimensions int, v vector);
CREATE TABLE
postgres=# INSERT INTO dimtest SELECT g, 1, ARRAY[1] FROM generate_series(1, 1000) g;
INSERT 0 1000
postgres=# INSERT INTO dimtest VALUES (0, 2, ARRAY[1,2]);
INSERT 0 1
postgres=# ANALYZE dimtest;
ANALYZE
postgres=# EXPLAIN SELECT * FROM dimtest WHERE dimensions = 2 AND v = ARRAY[1,2]::vector;
ERROR:  different vector dimensions 1 and 2
```
Without the ANALYZE, that works.

However, actually executing the query still fails. There is no guarantee on which order the quals are evaluated, and the executor chooses to evaluate the `v = ARRAY[1,2]::vector` qual before `dimensions = 2`. (Which actually surprises me a little, I thought the planner orders the quals by cost, and surely vector_eq() is marked with a higher cost than int4eq. I guess not. But there is no guarantee on the order anyway.) 

I don't know how to fix that without removing the dimension-checks, I'm afraid.

**Comment by ergo70** [comment]
I think that equality should simply return FALSE instead of an ERROR if the dimensions are different. At least, that ist the behavior of PostgreSQL ARRAY types, and it makes sense. Vectors of different size cannot be equal, regardless of their content. I was suprised too, that a separate dimension column is not evaluated first. My current workaround is pre-filtering in a CTE, but again, ARRAYS behave as expected when compared for equality, why shouldn't pgvector?

**Comment by ankane** [comment]
This change will be made in 0.7.0 (as mentioned above). I'd prefer not to make breaking changes in patch versions unless absolutely necessary, and I don't think this meets that criteria.

**Comment by ergo70** [comment]
No problem. Thank you for fixing it!

**Comment by ankane** [comment]
Pushed the second part of the change in the commit above, which will be included in 0.7.0 (#508). Thanks again for reporting.
